### PR TITLE
Wirelength Analyzer

### PR DESCRIPTION
**Pull request by zakn-amd**
Adds `wa.py`, a utility to compute critical-path wirelength of routed FPGAIF Physical Netlists.

### CODE REVIEW

**Review by eddieh-xlnx — APPROVED** [review]

File: wirelength_analyzer/README.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+       0|         | sitePin SLICE_X69Y92 AX
+       0|         | belPin  SLICE_X69Y92 AX AX
+       0|         | belPin  SLICE_X69Y92 CARRY8 AX
+        |       46| cell    u_calc/dropSpin/scattererReflector/squareRoot1_6/op__7_q_reg[63]_i_376__0
+<Truncated for brevity>
+```
+
+As seen above at this level of verbosity the `...` is replaced with the detailed
+path taken between two cells. The final verbosity level, level 0, prints the
+total wirelength and omits all other output. The second mode,
```
Comment (clavin-xlnx) [inline wirelength_analyzer/README.md:111]: Perhaps consider a paragraph break between explaining verbosity level 0 and `longest-single-net` mode.

File: wirelength_analyzer/test/test_wa.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        """
+        # build a graph of the Vivado route tree
+        vrt = ParseVivadoRouteTree(vrt_fname)
+        vrt.build_tree_from_net()
+
+        # compute the Longest Single Net
+        wa = WirelengthAnalyzer(phys_fname, 0)
+        wa.find_lsn()
+
+        # in single net mode the wirelenth anlayzer should return a path with
```
Comment (clavin-xlnx) [inline wirelength_analyzer/test/test_wa.py:87]:
```suggestion
        # in single net mode the wirelength analyzer should return a path with
```

File: wirelength_analyzer/wa.py [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+        if self.print_timing_commands:
+            tcl = self.vivado_timing_commands(cells_on_path)
+            print()
+            for cmd in tcl:
+                print(cmd)
+                print()
+
+    def vivado_timing_commands(self, cells_on_path):
+        """
+        Given a list of cell names return a list of Vivado TCL commands. The
```
Comment (clavin-xlnx) [inline wirelength_analyzer/wa.py:646]:
```suggestion
        Given a list of cell names return a list of Vivado Tcl commands. The
```

File: wirelength_analyzer/wa.py [hunk 4]
```diff
@@ -0,0 +1,731 @@
...
+        first command in the list reports timing through the list of cells, and
+        the second commmand in the list selects all of the cells in device
+        veiw.
+
+        Args:
+            cells_on_path: a list of cell names on a path
+        Returns:
+            a list of two Vivado TCL commands
```
Comment (clavin-xlnx) [inline wirelength_analyzer/wa.py:654]:
```suggestion
            a list of two Vivado Tcl commands
```

**Review by clavin-xlnx — COMMENTED** [review]
When adding the `-h` option, I get the following output:
```
usage: wa [-h] [-v VERBOSITY]
          [--mode {lsn,cp,longest-single-net,critical-path,both}]
          physical_netlist

Compute the longest wirelength in a routed FPGA Interchange Format Physical
Netlist

positional arguments:
  physical_netlist      FPGAIF Physical Netlist to process

optional arguments:
  -h, --help            show this help message and exit
  -v VERBOSITY, --verbosity VERBOSITY
                        verbosity
  --mode {lsn,cp,longest-single-net,critical-path,both}
                        Compute either the length of the longest single net or
                        the critical path or both
```

The `mode` usage is a bit strange as the other options follow a typical UNIX-style command line argument form, but mode does not.  At the least, I would change the output to make it obvious that there are really 2 (or 3 modes) rather than 5:

```
usage: wa [-h] [-v VERBOSITY]
          [--mode {lsn,cp,longest-single-net,critical-path,both}]
          physical_netlist

Compute the longest wirelength in a routed FPGA Interchange Format Physical
Netlist

positional arguments:
  physical_netlist      FPGAIF Physical Netlist to process

optional arguments:
  -h, --help            show this help message and exit
  -v VERBOSITY, --verbosity VERBOSITY
                        verbosity
  --mode MODE=
              {lsn, longest-single-net}
                    Compute the length of the longest single routed net 
              {cp, critical-path}
                    Compute the length of the critical path 
              {both}
                    Run both previous modes consecutively

```

File: wirelength_analyzer/Makefile [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# Copyright (C) 2023, Advanced Micro Devices, Inc.  All rights reserved.
+#
+# Author: Zak Nafziger, AMD
+#
+# SPDX-License-Identifier: MIT
+#
+
+ASSETS_URL = https://github.com/zakn-amd/fpga24_routing_contest/releases/download/test_assets_v1/test_assets.tar.gz
```
[the remark this answers is not in the document]
Comment (eddieh-xlnx) [reply wirelength_analyzer/Makefile:8]: I plan on updating the URL once we make the next release to which I can attach these assets.

File: wirelength_analyzer/README.md [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+For this competition a custom open-source wirelength analyzer, called `wa.py`
+has been written in Python and leverages the NetworkX graph library. The basic
+usage of this tool is:
+
+```
+python3 wa.py [-h] [-v VERBOSITY] [--mode {lsn,cp,longest-single-net,critical-path,both}] physical_netlist
+```
+
+The tool has three levels of verbosity and can operate in three different
+modes. The most basic form of operation is to simply supply the tool with the
```
[the remark this answers is not in the document]
Comment (zakn-amd) [reply wirelength_analyzer/README.md:11]: I restructured the help message here: https://github.com/Xilinx/fpga24_routing_contest/pull/25/commits/fbd6d60db73e450ba08c33702f7071ff2ae6adac. Unfortunately `argparse` doesn't seem to make it easy to add line breaks in a help message so the best I could get was:
```
usage: wa [-h] [-v VERBOSITY] [--mode MODE] physical_netlist

Compute the longest wirelength in a routed FPGA Interchange Format Physical Netlist

positional arguments:
  physical_netlist      FPGAIF Physical Netlist to process

options:
  -h, --help            show this help message and exit
  -v VERBOSITY, --verbosity VERBOSITY
                        verbosity
  --mode MODE           MODE={lsn, longest-single-net} -- compute the length of the longest single routed net. MODE={cp,
                        critical-path} -- compute the length of the critical path. MODE={both} -- run both previous modes
                        consecutively.

```

File: wirelength_analyzer/README.md [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+seen above this produces a table describing the path with the longest
+wirelength. The header of the table includes the total wirelength of
+the critical path, in the example above this is 663. All wirelength numbers
+reported by `wa.py` are dimensionless scores, for further detail about how
+these scores are computed and a description of the wirelength analyzer's
+algorithm see the [Scoring Criteria page](https://xilinx.github.io/fpga24_routing_contest/score.html).
+Each subsequent line in the table corresponds to one or more FPGA Interchange
+Format routeSegments, or combinatorial logic cells. The first column contains
+the wirelength of the routeSegments. The second column contains the running
+total wirelength and is indicated at every cell along the path. The Final
```
[the remark this answers is not in the document]
Comment (zakn-amd) [reply wirelength_analyzer/README.md]: Changed the wording here to mention the name of each column: https://github.com/Xilinx/fpga24_routing_contest/pull/25/commits/3b297010d0544da87bdf7dd0608dcd0c62404973

File: wirelength_analyzer/README.md (hunk 1 above)
Comment (zakn-amd) [reply wirelength_analyzer/README.md]: added a paragraph break: https://github.com/Xilinx/fpga24_routing_contest/pull/25/commits/3b297010d0544da87bdf7dd0608dcd0c62404973

**Comment by eddieh-xlnx** [comment]
Now the help message looks like:
```
$ python3 wirelength_analyzer/wa.py -h
usage: wa [-h] [-v VERBOSITY] [--mode MODE] physical_netlist

Compute the longest wirelength in a routed FPGA Interchange Format Physical
Netlist

positional arguments:
  physical_netlist      FPGAIF Physical Netlist to process

optional arguments:
  -h, --help            show this help message and exit
  -v VERBOSITY, --verbosity VERBOSITY
                        output verbosity level
  --mode MODE           MODE={lsn, longest-single-net} -- compute the length
                        of the longest single routed net. MODE={cp, critical-
                        path} -- compute the length of the critical path.
                        MODE={both} -- run both previous modes consecutively.
```

**Comment by eddieh-xlnx** [comment]
Now:
```
$ python3 wirelength_analyzer/wa.py -h
usage: wa [-h] [-v VERBOSITY] [--mode MODE] physical_netlist

Compute the longest wirelength in a routed FPGA Interchange Format Physical Netlist

positional arguments:
  physical_netlist      FPGAIF Physical Netlist to process

optional arguments:
  -h, --help            show this help message and exit
  -v VERBOSITY, --verbosity VERBOSITY
                        output verbosity level
  --mode MODE           MODE is 'lsn' or 'longest-single-net' -- compute the length of the longest single routed net.
                        MODE is 'cp' or 'critical-path' -- compute the length of the critical path.
                        MODE is 'both' -- run both previous modes consecutively.
```

**Pull request merged**